### PR TITLE
refactor: updates spectral and the rhoas guidelines to latest release

### PR DESCRIPTION
**Pull request by dimakis**
updates spectral and the rhoas guidelines to latest release

## Description

Bumps Spectral Linter and RHOAS Guidelines versions.
 
## Verification Steps

1. Check out the branch
2. Run `make openapi/spec/validate`
The make file target should look for the older version, if it exists it will remove it then install and link the new executable. 

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

File: .spectral.yaml [hunk 1]
```diff
@@ -1,6 +1,11 @@
-extends: '@rhoas/spectral-ruleset'
+extends: 'bin/rhoasapi-installation/node_modules/@rhoas/spectral-ruleset/ruleset.yaml'
 rules:
   no-$ref-siblings: off
   rhoas-response-media-type: off
   oas3-valid-schema-example: off
+  rhoas-external-$ref: off
```
Comment (machi1990) [inline .spectral.yaml:6]: Why is this turned off?

File: Makefile [hunk 2]
```diff
@@ -146,24 +146,26 @@ SPECTRAL ?= ${LOCAL_BIN_PATH}/spectral
...
+    if ! [ -L ${LOCAL_BIN_PATH}/rhoasapi ]; then ln -s ${LOCAL_BIN_PATH}/rhoasapi-installation/node_modules/@rhoas/spectral-ruleset/rhoasapi ${LOCAL_BIN_PATH}/rhoasapi; fi ;\
 	}
 endif
 
+
+.PHONY: openapi/spec/validate
 openapi/spec/validate: specinstall
-	$(SPECTRAL) lint openapi/kas-fleet-manager.yaml
-	$(SPECTRAL) lint -s 'rhoas-external-$$ref' -s rhoas-list-schema -s rhoas-object-schema -s rhoas-error-schema openapi/kas-fleet-manager-private-admin.yaml
+	@npx ${LOCAL_BIN_PATH}/rhoasapi lint openapi/kas-fleet-manager.yaml openapi/kas-fleet-manager-private-admin.yaml
```
Comment (machi1990) [inline Makefile:167]: why `npx`?

File: Makefile [hunk 3]
```diff
@@ -146,24 +146,26 @@ SPECTRAL ?= ${LOCAL_BIN_PATH}/spectral
...
-	cd ${LOCAL_BIN_PATH} ;\
-	${NPM} install --prefix ${LOCAL_BIN_PATH}/spectral-installation @stoplight/spectral@5.9.2 ;\
-	${NPM} i --prefix ${LOCAL_BIN_PATH}/spectral-installation @rhoas/spectral-ruleset@0.1.4 ;\
-	ln -s spectral-installation/node_modules/.bin/spectral spectral ;\
+    set -e ;\
+    mkdir -p ${LOCAL_BIN_PATH} ;\
+    mkdir -p ${LOCAL_BIN_PATH}/rhoasapi-installation ;\
+	if [ -d ${LOCAL_BIN_PATH}/spectral-installation ]; then rm -r ${LOCAL_BIN_PATH}/spectral-installation; fi ;\
+	if [ -L ${LOCAL_BIN_PATH}/spectral ]; then rm ${LOCAL_BIN_PATH}/spectral; fi ;\
+     ${NPM} i --prefix ${LOCAL_BIN_PATH}/rhoasapi-installation @rhoas/spectral-ruleset@0.3.0-dev6 ;\
```
Comment (machi1990) [inline Makefile:159]: This is using a dev ruleset `0.3.0-dev6`, what does that mean in terms of stability of the ruleset? Shouldn't we be using the stable-ish release instead?

File: .spectral.yaml [hunk 4]
```diff
@@ -1,6 +1,11 @@
-extends: '@rhoas/spectral-ruleset'
+extends: 'bin/rhoasapi-installation/node_modules/@rhoas/spectral-ruleset/ruleset.yaml'
```
Comment (machi1990) [inline .spectral.yaml:1]: Why this change?

File: Makefile [hunk 5]
```diff
@@ -146,24 +146,26 @@ SPECTRAL ?= ${LOCAL_BIN_PATH}/spectral
...
-ifeq (, $(shell which ${LOCAL_BIN_PATH}/spectral 2> /dev/null))
+ifeq (, $(shell which ${LOCAL_BIN_PATH}/rhoasapi 2> /dev/null))
 	@{ \
-	set -e ;\
-	mkdir -p ${LOCAL_BIN_PATH} ;\
-	mkdir -p ${LOCAL_BIN_PATH}/spectral-installation ;\
-	cd ${LOCAL_BIN_PATH} ;\
-	${NPM} install --prefix ${LOCAL_BIN_PATH}/spectral-installation @stoplight/spectral@5.9.2 ;\
-	${NPM} i --prefix ${LOCAL_BIN_PATH}/spectral-installation @rhoas/spectral-ruleset@0.1.4 ;\
-	ln -s spectral-installation/node_modules/.bin/spectral spectral ;\
```
Comment (machi1990) [inline Makefile:160]: SPECTRAL is also used in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/feb9f25d99047f86e0151bb5d595e6b3c197eb2b/Makefile#L299 and I am wondering at what point we are installing it after this removal?

File: Makefile [hunk 6]
```diff
@@ -146,24 +146,26 @@ SPECTRAL ?= ${LOCAL_BIN_PATH}/spectral
 NPM ?= "$(shell which npm)"
 specinstall:
 ifeq (, $(shell which ${NPM} 2> /dev/null))
-	@echo "npm is not available please install it to be able to install spectral"
-	exit 1
+       @echo "npm is not available please install it to be able to install spectral"
+       exit 1
 endif
-ifeq (, $(shell which ${LOCAL_BIN_PATH}/spectral 2> /dev/null))
+ifeq (, $(shell which ${LOCAL_BIN_PATH}/rhoasapi 2> /dev/null))
```
Comment (miguelsorianod) [inline Makefile:152]: Could you give more context on what is `rhoasapi`, the difference/s with `spectral` and why has it been changed?

File: .spectral.yaml (hunk 1 above)
Comment (dimakis) [reply .spectral.yaml:6]: Updated some of the rules to give warnings instead of error out, the reasoning is because the openapi generator can now deal with external refs in OAS files.

File: Makefile (hunk 2 above)
Comment (dimakis) [reply Makefile:167]: runs the local executable

Comment (miguelsorianod) [inline Makefile:167]: Can we rename the SPECTRAL Makefile variable to RHOASAPI or SPECTRAL_RHOASAPI (whatever is more appropriate depending on what the binary is)?

Comment (miguelsorianod) [inline Makefile:167]: Why is npx needed?

File: Makefile (hunk 3 above)
Comment (dimakis) [reply Makefile:159]: this will be very quickly bumped to a full release :) with very little changing. We decided to see how painful / painless updating all of the apis would be

File: Makefile (hunk 2 above)
Comment (machi1990) [reply Makefile:167]: Yes, I know. But is it needed? See previously how the local executable was run.

File: .spectral.yaml (hunk 4 above)
Comment (dimakis) [reply .spectral.yaml:1]: newest spectral version will not locate the ruleset without being expressly told where it is. it is indeed uglier but linter will not find the ruleset otherwise

File: .spectral.yaml (hunk 1 above)
Comment (machi1990) [reply .spectral.yaml:6]: I might be misunderstanding something, if the rule is set to `off` won't it be ignored completely?

File: Makefile (hunk 2 above)
Comment (dimakis) [reply Makefile:167]: its not necessary to be honest. I've just tested without I can remove it. I used it in testing to make sure I used the local executable

Comment (dimakis) [reply Makefile:167]: it is not. I've just testing it without. I'll remove it. I used it during testing

File: .spectral.yaml (hunk 4 above)
Comment (machi1990) [reply .spectral.yaml:1]: :-( this sounds like a bug more than a feature

File: Makefile (hunk 3 above)
Comment (machi1990) [reply Makefile:159]: It'll be awesome to bump to a full release and not leave it a dev. Or have a followup issue where we track this to avoid forgetting about it ;-)

File: Makefile (hunk 5 above)
Comment (dimakis) [reply Makefile:160]: this is a great catch. I didn't know about this. I'll fix this

File: Makefile (hunk 2 above)
Comment (dimakis) [reply Makefile:167]: indeed, I'll get to that now

File: Makefile (hunk 6 above)
Comment (dimakis) [reply Makefile:152]: `rhoaspi` is essentially spectral binary thats been wrapped up in the rhoas guidelines package

File: .spectral.yaml (hunk 1 above)
Comment (dimakis) [reply .spectral.yaml:6]: yes it will.

File: .spectral.yaml (hunk 4 above)
Comment (dimakis) [reply .spectral.yaml:1]: I genuinely do not know what the reason is for this. it appears to be just this repo that has the issue. want me to create a link to it instead?

File: .spectral.yaml (hunk 1 above)
Comment (machi1990) [reply .spectral.yaml:6]: Thanks. Then I guess what we want is to always enable this so that we get the warning message we need. 
i.e the line `rhoas-external-$ref: off` will have to be removed.

File: Makefile (hunk 6 above)
Comment (machi1990) [reply Makefile:152]: > `rhoaspi` is essentially spectral binary thats been wrapped up in the rhoas guidelines package

Does the wrapping assumes that `spectral` cli is already installed?
I am getting 
```
bin/rhoasapi: line 5: spectral: command not found
npm WARN using --force I sure hope you know what you are doing.
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /usr/local/lib/node_modules/.staging
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/.staging'
npm ERR!  [OperationalError: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/.staging'] {
npm ERR!   cause: [Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/.staging'] {
npm ERR!     errno: -13,
npm ERR!     code: 'EACCES',
npm ERR!     syscall: 'mkdir',
npm ERR!     path: '/usr/local/lib/node_modules/.staging'
npm ERR!   },
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib/node_modules/.staging',
npm ERR!   parent: '@rollup/pluginutils'
npm ERR! }
npm ERR! 
```

when I run the change locally.

File: .spectral.yaml (hunk 4 above)
Comment (machi1990) [reply .spectral.yaml:1]: Yes. A link to the issue will be fine. Thank you so much.

File: Makefile (hunk 6 above)
Comment (dimakis) [reply Makefile:152]: no it doesn't. they are both bundled together. So that is the output when you run `make openapi/spec/validate`?

Comment (machi1990) [reply Makefile:152]: Yes. That's what I got when I run `make openapi/spec/validate`

**Comment by jackdelahunt** [comment]
Went through verification steps and got expected output.

LGTM :+1:

File: Makefile (hunk 6 above)
Comment (dimakis) [reply Makefile:152]: I've pushed some new changes, only a new directory is now created (this just takes the place of the old `spectral-installation` dir.  This is done to facilitate the easy clean up of existing spectral and RHOAS Guidelines which may be on dev machines.

running `make openapi/spec/validate` should now clean up, install new versions and lint.

**Review by machi1990 — COMMENTED** [review]
Great work @dimakis 

Validation works, seeing these warnings which are to be expected right?
```
make openapi/spec/validate

/home/manyanda/Documents/Projects/Kafka/managed-services-api/openapi/kas-fleet-manager-private-admin.yaml
 204:11  warning  rhoas-error-schema   Missing required schema object "Error"            components.schemas
 204:11  warning  rhoas-list-schema    Missing required schema object "List"             components.schemas
 204:11  warning  rhoas-object-schema  Missing required schema object "ObjectReference"  components.schemas

/home/manyanda/Documents/Projects/Kafka/managed-services-api/openapi/kas-fleet-manager.yaml
 139:23  warning  oas3-valid-media-example  "example" property type must be object            paths./api/kafkas_mgmt/v1/kafkas/{id}.delete.responses[202].content.application/json.example
 257:16  warning  rhoas-operaton-id         "parameters.operationId" property must be truthy  paths./api/kafkas_mgmt/v1/kafkas/{id}.parameters
```

**Comment by dimakis** [comment]
> Great work @dimakis
> 
> Validation works, seeing these warnings which are to be expected right?
> 
> ```
> make openapi/spec/validate
> 
> /home/manyanda/Documents/Projects/Kafka/managed-services-api/openapi/kas-fleet-manager-private-admin.yaml
>  204:11  warning  rhoas-error-schema   Missing required schema object "Error"            components.schemas
>  204:11  warning  rhoas-list-schema    Missing required schema object "List"             components.schemas
>  204:11  warning  rhoas-object-schema  Missing required schema object "ObjectReference"  components.schemas
> 
> /home/manyanda/Documents/Projects/Kafka/managed-services-api/openapi/kas-fleet-manager.yaml
>  139:23  warning  oas3-valid-media-example  "example" property type must be object            paths./api/kafkas_mgmt/v1/kafkas/{id}.delete.responses[202].content.application/json.example
>  257:16  warning  rhoas-operaton-id         "parameters.operationId" property must be truthy  paths./api/kafkas_mgmt/v1/kafkas/{id}.parameters
> ```

Yes that's exactly what we should be seeing. great 😄

**Review by machi1990 — COMMENTED** [review]
The PR includes commits from other PR, I suspect a rebase going wrong.
Can you double check? 
Other than that the change lgtm

**Comment by machi1990** [comment]
Can we also update the https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager#prerequisites to update the NodeJS version. I think v14 is now the minimum required one.

**Comment by dimakis** [comment]
> The PR includes commits from other PR, I suspect a rebase going wrong.
> Can you double check?
> Other than that the change lgtm

yea, rebase gone wrong. should be ok now. 

I've also bumped the minimum required version of node in the readme. great catch @machi1990

File: Makefile (hunk 2 above)
Comment (JameelB) [reply Makefile]: I still see this as `SPECTRAL` in the makefile @dimakis

**Comment by JameelB** [comment]
Thanks for this @dimakis. I've verified the installation and it works as expected. Just one comment there based on @miguelsorianod previous review.

File: Makefile (hunk 2 above)
Comment (dimakis) [reply Makefile]: It is spectral though, the spectral binary from the installation of `spectral-cli`. As spectral is a needed dependency, it will be installed regardless. I'm leaving the install in to make it easier to read.

**Comment by dimakis** [comment]
Thanks @JameelB, question, do ye prefer to have the versions locked in? i.e the way that spectral cli and rhoas guidelines are defined or just leave the tag to latest?

**Comment by JameelB** [comment]
> Thanks @JameelB, question, do ye prefer to have the versions locked in? i.e the way that spectral cli and rhoas guidelines are defined or just leave the tag to latest?

As per our discussion, let's leave this to use a specific tag rather than latest.

**Review by JameelB — APPROVED** [review]
